File: frontend/src/endpoints/subtitle.ts
import { type ApiResponse, apiUrl, delete_, get, post, put } from "@/endpoints/request";
import type { VideoId } from "@/endpoints/video";
import type { Id } from "@/types/id";

export type SubtitleId = Id;

export interface SubtitleInfo {
  subtitlesExtracted: boolean;
}

export function getSubtitleInfo(videoId: VideoId): Promise<ApiResponse<void, SubtitleInfo>> {
  return get(`/videos/${videoId}/subtitles/info`);
}

export interface Subtitle {
  id: Id;
  name: string;
}

export function extractSubtitles(videoId: VideoId): Promise<ApiResponse<void, Subtitle[]>> {
  return post(`/videos/${videoId}/subtitles/extract`);
}

export function getAllSubtitles(videoId: VideoId): Promise<ApiResponse<void, Subtitle[]>> {
  return get(`/videos/${videoId}/subtitles`);
}

export function subtitleContentUrl(videoId: VideoId, subtitleId: SubtitleId): string {
  return apiUrl(`/videos/${videoId}/subtitles/${subtitleId}/content`);
}

export function getSubtitleContent(
  videoId: VideoId,
  subtitleId: SubtitleId
): Promise<ApiResponse<void, string>> {
  return get(`/videos/${videoId}/subtitles/${subtitleId}/content`);
}

export function createSubtitle(
  videoId: VideoId,
  fileContent = ""
): Promise<ApiResponse<void, Subtitle>> {
  return post(`/videos/${videoId}/subtitles`, fileContent);
}

export interface UpdateSubtitleInfo {
  name: string;
  content: string;
}

export function updateSubtitle(
  videoId: VideoId,
  subtitleId: SubtitleId,
  update: UpdateSubtitleInfo
): Promise<ApiResponse<UpdateSubtitleInfo, void>> {
  return put(`/videos/${videoId}/subtitles/${subtitleId}`, update);
}

export function deleteSubtitle(
  videoId: VideoId,
  subtitleId: SubtitleId
): Promise<ApiResponse<void, void>> {
  return delete_(`/videos/${videoId}/subtitles/${subtitleId}`);
}

File: backend/models/subtitle.go
package models

type Subtitle struct {
	Base

	VideoId Id `gorm:"not null"`
	Video   Video

	Name string `gorm:"not null; type:text"`
}

func CreateSubtitle(id Id, videoId Id, Name string) (*Subtitle, error) {
	subtitle := &Subtitle{
		Base:    Base{Id: id},
		VideoId: videoId,
		Name:    Name,
	}
	err := db.Create(subtitle).Error
	return subtitle, err
}

func GetAllSubtitles(video Id) ([]Subtitle, error) {
	var subtitles []Subtitle
	err := db.Order("name").Find(&subtitles, &Subtitle{VideoId: video}).Error
	return subtitles, err
}

func GetSubtitle(id Id) (*Subtitle, error) {
	subtitle := &Subtitle{}
	err := db.Find(subtitle, &Subtitle{Base: Base{Id: id}}).Error
	return subtitle, err
}

func (subtitle *Subtitle) Save() error {
	return db.Save(subtitle).Error
}

func (subtitle *Subtitle) Delete() error {
	return db.Delete(&Subtitle{Base: Base{Id: subtitle.Id}}).Error
}
